### PR TITLE
WIP: add benchmark for cluster_qr

### PR DESCRIPTION
**Pull request by victorminden**
This is a WIP incomplete PR to capture the state of initial patches to `lobpcg:patch-1` in the interest of ultimate inclusion in https://github.com/scikit-learn/scikit-learn/pull/21148.

### CODE REVIEW

File: benchmarks/bench_plot_spectral.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        .fit(S)
+        .labels_
+    )
+    t1 = time() - t0
+    score = adjusted_rand_score(y, labels)
+    
+    return t1, score
+
+
+def run_benchmark():
```
Comment (victorminden) [inline benchmarks/bench_plot_spectral.py:38]: This entire file is WIP and just basically cut-and-paste from the test so far.  In beginning to look at a basic benchmark here, it became clear (IMO) that users will see no practical runtime difference between these methods because the `spectral_clustering` call includes the eigensolves.  I expect that the eigensolves will dominate the computation for any one of these methods, unless perhaps the number of clusters is comparable to the number of data points, and so there should be no appreciable runtime difference in the user-visible API I think.

Therefore, I don't think it makes sense to benchmark runtime.  WDYT?  @lobpcg

**Comment by lobpcg** [comment]
You and I know that labeling itself takes little time compared to the eigensolves, but  other people may not. The same actually should be true for memory. So we should report time and memory to make the point that it makes no real difference.

**Comment by lobpcg** [comment]
@victorminden we do not need to make the benchmark as a part of the branch yet, until it is finalized and follows all the standards. At this point a local script in any format is fine since we only want to see and report preliminary benchmark data. Maybe do not create PR until the more-or-less final version of the benchmark is ready for PR?

**Comment by victorminden** [comment]
To be clear, this is a draft PR to your branch on your personal fork, not
anything else.  I'm not going to do any development on this outside of a
git branch, and I needed to discuss with you the proposed changes and
content of that branch, so a draft PR seemed appropriate..  If you prefer
not to see a draft PR open to your fork, feel free to close it, of course.

Best,

Victor

--
Victor L. Minden
***@***.***

File: benchmarks/bench_plot_spectral.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+from sklearn.cluster import SpectralClustering
+from sklearn.metrics import adjusted_rand_score
+from sklearn.metrics.pairwise import kernel_metrics, rbf_kernel
+from sklearn.datasets import make_blobs
+
+
+def get_data():
+    for size in (50, 100, 500, 1000):#, 5000):
+        X_file = csv.reader(open(f"simulated_blockmodel_graph_{size}_nodes.tsv"),
```
[the remark this answers is not in the document]
Comment (lobpcg) [reply benchmarks/bench_plot_spectral.py:19]: It's possible that spectral clustering works so well here that specific labeling method is not so important.  Please look below on that page and grab examples with larger spread.

**Comment by lobpcg** [comment]
Keeping your benchmark on the branch is a good idea allowing all to see it, if you don't mind a bit of extra work of updating it. 

On examples, the most difficult ones there are High Block Overlap and High Block Size Variation to the point that the ground truth may not in some examples give the best clusters. High block size requires using the normalized Laplacian, so we need to double check if this is the default in this code.

**Comment by victorminden** [comment]
Yes, the code uses the normalized Laplacian.  I added a few cases per your suggestions.  `kmeans` isn't great on one of them but otherwise everything is still pretty comparable.  Note that 5000/high variability seems to not be a connected graph, so I did not do it.  Anything bigger than this is not going to run on my laptop.

Feel free to play with this, add new benchmark cases, as you like.  I won't have much time to look at this further this week, but this basic skeleton seems fine for benchmarking based on the other sklearn examples.

```
Test static_highOverlap_lowBlockSizeVar of size (1000, 1000) with 11 clusters
kmeans    : 0.947 (0.258 sec., 117.777 MB)
discretize: 0.968 (0.193 sec., 119.695 MB)
cluster_qr: 0.961 (0.193 sec., 120.055 MB)


Test static_highOverlap_highBlockSizeVar of size (1000, 1000) with 11 clusters
kmeans    : 0.866 (0.229 sec., 120.340 MB)
discretize: 0.971 (0.191 sec., 120.535 MB)
cluster_qr: 0.957 (0.187 sec., 120.594 MB)


Test static_highOverlap_lowBlockSizeVar of size (5000, 5000) with 19 clusters
kmeans    : 0.991 (17.415 sec., 403.094 MB)
discretize: 0.997 (17.209 sec., 403.094 MB)
cluster_qr: 0.995 (17.301 sec., 403.523 MB)
```

**Comment by lobpcg** [comment]
@victorminden fantastic, thanks! The results you report look plausible. I'll see if I could run your benchmark. I actually do not have even a local copy of my own branch, doing all editing over the web interface this time so far.

One of their posted examples was broken, probably the one that you have also found.

**Comment by lobpcg** [comment]
It is a better practice to limit PRs to focused changes.

Suggesting small changes in my branch is more effective to do using "review" in my branch rather than via PR. 

Anyway, I have incorporated all your proposed changes in  in my branch, exept for the benchmarking itself. Could you please review when you find time?

**Comment by lobpcg** [comment]
The problem with high-overlap tests is that the true partition is not necessarily "optimal," as a large random perturbation of a true value may move the point closer to a different true value.

May be we should try again low-overlap tests, where the ground truth is accurate, but limit the tolerance  of the eigenvectors to make labeling work harder.

**Comment by victorminden** [comment]
I disagree, I am not a formal reviewer of your PR and if I litter it with
comments it is just going to obfuscate the discussion on your PR with
actual reviewers.

I factored my WIP PR into individual focused commits that you can
incorporate or not. Of course this sounds more difficult if you are using
solely the web interface, but I couldn't say.

Mobile

**Comment by victorminden** [comment]
As you like -- as I said feel free to play with whatever test cases you
want.  Trading off accuracy for runtime is fine with me for generating some
more interesting benchmarks, but it's not something that I personally am
going to have time for in the near-term.

Sklearn has instructions for how to install your local sklearn source as an
editable library, that will let you get started with running things (takes
a few minutes to cythonize things but otherwise easy)

Mobile
